Services/UI/OJS.Services.Ui.Business/Implementations/ContestsBusinessService.cs

```diff
@@ -178,15 +177,53 @@ public async Task<ContestDetailsServiceModel> GetContestDetails(int id)
             registerModel.RequirePassword = ShouldRequirePassword(contest!, participant!, isOfficial);
             registerModel.ParticipantId = participant?.Id;
             registerModel.IsRegisteredSuccessfully = participant != null;
-            registerModel.ShouldConfirmParticipation = contest!.IsOnlineExam &&
-                                                       isOfficial &&
-                                                       participant == null &&
-                                                       !userIsAdminOrLecturerInContest;
+            registerModel.ShouldConfirmParticipation = ShouldConfirmParticipation(participant, isOfficial, contest!.IsOnlineExam, userIsAdminOrLecturerInContest);
+
+            return registerModel;
+        }
+
+        public async Task<bool> ValidateUserRegistrationForContest(
+            int id,
+            string? password,
+            bool? hasConfirmedParticipation,
+            bool isOfficial)
+        {
+            var contest = this.contestsData
+                .GetByIdQuery(id)
+                .Include(c => c.Category)
+                .Include(c => c.ProblemGroups)
+                .ThenInclude(pg => pg.Problems)
```

Comment:
Why do we need the Problems to validate the user registration for a contest?

Reply:
We need them for `AssignRandomProblemsToParticipant` logic.

Reply:
By looking the name it seems the method does only some kind of validation, but not actual registration.